packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -29,64 +38,198 @@ class UpgradeCommand extends FlutterCommand {
+          'changes. If it is okay to remove local changes, then re-run this '
+          'command with --force.'
+        );
+      }
+    }
+    final String stashName = await maybeStash(gitTagVersion);
+    // If we're on a hotfix version, that means we likely have a different
+    // history than the tags pulled from master. We need to reset to the
+    // tag this hotfix was branched from.
+    if (gitTagVersion.hotfix != null) {
+      await unapplyHotfix(gitTagVersion);
+    }
+    await upgradeChannel(flutterVersion);
+    await attemptFastForward();
+    await applyStash(stashName);
+    await precacheArtifacts();
+    await updatePackages(flutterVersion);
+    await runDoctor();
+    return null;
+  }
+
+  /// Check if there is an upstream repository configured.
+  ///
+  /// Exits tool if there is no upstream.
+  Future<void> verifyUpstreamConfigured() async {
     try {
       await runCheckedAsync(<String>[
         'git', 'rev-parse', '@{u}',
       ], workingDirectory: Cache.flutterRoot);
     } catch (e) {
-      throwToolExit('Unable to upgrade Flutter: no upstream repository configured.');
+      throwToolExit(
+        'Unable to upgrade Flutter: no origin repository configured. '
+        'Run \'git remote add origin '
+        'https://github.com/flutter/flutter\' in ${Cache.flutterRoot}',
+      );
+    }
+  }
+
+  /// Attempt to stash any local changes.
+  ///
+  /// Returns the stash name if any changes were stashed. Exits tool if
+  /// `git stash` returns a non-zero exit code.
+  Future<String> maybeStash(GitTagVersion gitTagVersion) async {
+    try {
+      final RunResult runResult = await runCheckedAsync(<String>[
+        'git', 'status', '-s',
+      ]);
+      // If there are no local changes, skip the stash.
+      if (runResult.stdout.trim().isEmpty) {
+        return null;
+      }
+    } catch (e) {
+      throwToolExit('Failed to check git status: $e');
+    }
+    final String stashName = 'flutter-upgrade-from-v${gitTagVersion.x}.${gitTagVersion.y}.${gitTagVersion.z}';
+    try {
+      final RunResult runResult = await runCheckedAsync(<String>[
+        'git', 'stash', 'push', '-m', stashName
+      ]);
+      // output message will contain stash name if any changes were stashed..
+      if (runResult.stdout.contains(stashName)) {
+        return stashName;
+      }
+    } catch (e) {
+      throwToolExit('Failed to stash local changes: $e');
     }
+    return null;
+  }
 
-    final FlutterVersion flutterVersion = FlutterVersion.instance;
+  /// Attempts to reset to the un-hotfixed version. This should restore the
+  /// history to something that is compatible with the regular upgrade
+  /// process.
+  Future<void> unapplyHotfix(GitTagVersion gitTagVersion) async {
+    final RunResult runResult = await runCheckedAsync(<String>[
+      'git', 'reset', '--hard', 'v${gitTagVersion.x}.${gitTagVersion.y}.${gitTagVersion.z}',
+    ]);
```

Comment:
I think it would be much safer to always do this. Otherwise we'll have code that we're not really testing very often which will bitrot and then when we realise our mistake it'll be extremely fatal.

We should be more paranoid of this code than anything else in the entire project. Write this with the most in-depth defensive programming you've ever used. Minimise conditionals and branches as much as possible. Be as flexible with the environment as you can be.

Reply:
Lets drop the stashing logic then. We've already seen cases where that particular stash command was not supported by the version of git on cirrus, I don't trust taking that into production